### DISCUSSION TITLE
PyScript builtins and stdlib

### DISCUSSION THREAD

**Discussion by fpliger** [pyscript/pyscript#1687] [Proposals]
I've been working on `pydom`, docs, and examples, while also trying to ask around out of the core maintainers group and understand what feels natural and "pythonic".

This morning I've been pairing with @tedpatrick and @mchilvers on what is the most natural place to have `pydom` and how/where to bring `when` and `HTML` back to PyScript.

I know we are ready to release but these things are really important (I wouldn't want for us to change something right now, break things and then revert back right after).

I'd like to discuss what makes me (and I think @tedpatrick  and @mchilvers as well, but I don't want to speak for others) feel a bit off....

1. We removed `when` and `HTML`. We should bring these back.
2. We changed `display` from being a builtin already loaded into the global namespace and made it explicit as an `import`
3. We need to understand where `pydom` and all other Python APIs for the web should live.

Here's where @mchilvers , @tedpatrick and myself ended up landing after some really good discussions pairing this morning.

1. We should bring `when` and `HTML` back before releasing (this should be fairly simple as they can easily be ported)
2. `display` seems really part of what PyScript brings on top of Python and makes a lot of sense for it not need to be imported, just like `print`. The argument here is that as `print` is the way the Python language provide a way for users to print an output [to the console] and it's just part of the language (and you don't need to import it). If Python was created with an interface to the browser right away and with the ability to output to the page vs to the console , would `display` be a keyword as `print`? Likely yes. 
3. A similar argument could be made for `when` and `HTML`, even if there's probably a weaker argument here
4. To steal Ted's words "It's easier to give than take back" . Basically, keep them as they are in the last release of classic (you don't need to import from anywhere now)
5. `pydom` and the other pythonic APIs to the web feel like they should belong closed to the `pyscript` package namespace while we keep something else, like `pyweb`(?) that can be used as a namespace to collect all the 3rd party extensions that add to the platform, making it easier to find them. So, for instance, a `pip install pyscript-shoelace` cmd could install a new UI widgets library powered by Shoelace under `pyweb.shoelace`
6. To be explicit about the above, the proposal is that for now we only add `pydom` under `pyscript.com`

This shouldn't mean any massive internal changes but rather just organizing scopes and namespaces. If that seems more laborious than that, please bring it out.

Let's try to clarify this before we release.

**Comment by JeffersGlass** [comment]
I think the argument for `display` as the "browser equivalent of print" (big ole scare quotes there) is valid, and I think the user should have to import `when` and `HTML` themselves. I see what you mean that those latter two _could_ be imported for the user... but then if we have other event-related or formatting-related functions in the future, are we importing _those_? I think not. I vote that we draw a line under `display`, import it by default, and nothing further.

One quick note on (5): You do have to import `when` in the last version of PyScript Classic. The only names we imported 'for the user' were: `js`, `pyscript`, `Element`, `display`, and `HTML`. [See here](https://github.com/pyscript/pyscript/blob/848d77b1c29d4145a3438b832acd4b3b9bf32951/pyscriptjs/src/remote_interpreter.ts#L150-L154).

**Reply by fpliger** [reply]
Did you mean (3)? But, yeah.. thanks for pointing it out!

**Reply by fpliger** [reply]
Just to make my position explicit, I think `display` is the `print` for the web and should be a built-in function. I think `when` should be as well but not as strong about it as with `display`. I'm quite neutral about `HTML`.

**Comment by JeffersGlass** [comment]
It would be good to document whatever we do - the question "Why can I just use Element/HTML/display without importing it?" has come up on the subreddit [a couple times](https://www.reddit.com/r/PyScript/comments/14hb63r/import_element_necessary_for_dom_access/).

**Reply by WebReflection** [reply]
People surprised we pollute globals is a strong indication we shouldn’t, imho

**Reply by fpliger** [reply]
> It would be good to document whatever we do - the question "Why can I just use Element/HTML/display without importing it?" has come up on the subreddit [a couple times](https://www.reddit.com/r/PyScript/comments/14hb63r/import_element_necessary_for_dom_access/).

Absolutely. Our docs need a lot of love :) Trying to work on it.

> People surprised we pollute globals is a strong indication we shouldn’t, imho

That's often a good argument but in this context, I've heard people be surprised about both sides of the spectrum. That's especially true when starting to consider the person's background (Python dev? web dev? novice? etc...). We should listen to the community but also provide governance and consider the different personas in our base when making decisions like these. Especially when the data pool collected from the community is indirect and loose.

**Reply by WebReflection** [reply]
I don’t think it’s a good practice in Python to pollute the global scope and I haven’t heard any complain about the lack of global pollution in PyScript Next. Do you have any link that states otherwise?

**Reply by fpliger** [reply]
> I don’t think it’s a good practice in Python to pollute the global scope and I haven’t heard any complain about the lack of global pollution in PyScript Next. Do you have any link that states otherwise?

I think we should align on what each other mean with `pollution` or `leak`

I read `leak` or `pollution` as objects that end up in `globals` unintentionally. If they are intentionally placed in `globals`, like Python build-ins, they are not `leaking` but rather just intentionally made available by design.

**Reply by WebReflection** [reply]
for clarification sake, you are right *pollution* is strictly more correct (we pollute on purpose either builtins or globals) but leak is also what I meant because by polluting we inevitably allocate memory to a reference that *might* not be needed by the program and it's there for no purpose (even if reachable). Not all examples are based on *display*, neither not all integration tests, hence I've stretched the meaning of leak, but that's what it looks like when nobody meant to use `display` or anything else ... they still sit on the virtual FS but not as reference in the running code. With Antonio suggestion though, I think this clarification is not needed as it's not a leak, rather an intentional default help we provide to our users, to which I'm more than happy about.

**Comment by WebReflection** [comment]
I like the fact we spent time to be sure no global leaks happen and then we want global leaks to happen 😅

https://github.com/pyscript/pyscript/issues/1651

in JS is just a bad practice to set anything global since ESM exist and I believe the same can be stated for Python.

`display` as global makes it impossible to have a `display` module and I disagree it would be the Python equivalent for `print`.

We already pollute globals with our own module but that’s sensible.

What concerns me the most is that in the IoT world `display` has an hardware meaning too and us deciding to shadow that meaning doesn’t feel right to me.

having an `import pyscript` also brands the code + to leak display globally, or anything else, all anyone need is to add a tiny script on top of the page that `from pyscript import display, HTML, when` and call it a day.

Moreover, the display has an automatic target which mislead if assumed as global instead of being imported explicitly like current approved tests, MRs, and integration tests show.

I hence respectfully disagree on leaking anything from our lovely module because there are easy workarounds to leak globally anything but there won’t be workarounds if a display module exists in the future.

leak one thing or many is the same slippery slope but if nothing I’ve said here convinces you I hope at least we draw that line that one leak is already too much (what Jeff said).

**Reply by WebReflection** [reply]
Forgot to mention this:

> To steal Ted's words "It's easier to give than take back"

that's why we should offer PyScript Next with best practices applied (no default global ~~leaks~~ pollution) and eventually listen to the community if the community believes `display` should be globally available. Doing the other way around means we'll never be able to remove that `display` if we want to, or we need to, in the future, and as we agreed PyScript Next was the breaking change (in a way or another) I think it would be wise to start following best standard practices because it's easier to add later on than to remove later on without breaking everything.

My 2 cents, I wish @antocuni and @hoodmane would also express their thoughts here.

**Comment by fpliger** [comment]
> I like the fact we spent time to be sure no global leaks happen and then we want global leaks to happen 😅

That's a different thing. `display` being a [built-in function](https://docs.python.org/3/library/functions.html) is not a leak, it's an intentional choice because we choose to make it always available (just like `print` in Python or `console` in JS). 

> display as global makes it impossible to have a display module and I disagree it would be the Python equivalent for print.

Why we can't have already run `from pyscript import display`? 

> What concerns me the most is that in the IoT world display has an hardware meaning too and us deciding to shadow that meaning doesn’t feel right to me.

Can you provide a real use case around this?

> having an import pyscript also brands the code + to leak display globally, or anything else, all anyone need is to add a tiny script on top of the page that from pyscript import display, HTML, when and call it a day.

That could have been an argument for any `built-in` function in general. I wouldn't want for 90% of the PyScript apps to always start with `from pyscript import display`

Being a built-in is not a leak, it's an intentional exposure

**Reply by WebReflection** [reply]
It’s not a builtin… your link doesn’t have a display builtin so it’s an obtrusive move to me as we have a module and asking Python developers to use a module is the normality, deciding ourselves what is builtin and what isn’t is not. MicroPython doesn’t have it neither, and neither does PythonMonkey. Why do we feel so special about what should be a builtin?

**Reply by WebReflection** [reply]
> Why we can't have already run from pyscript import display?

if there is a display module users can import display as py_display and be more explicit, but I’m not sure I understood your question.

**Reply by WebReflection** [reply]
> Can you provide a real use case around this?

display is already a pip module, that’s what I’m talking about: https://pypi.org/project/display/

we can’t make up builtins, imho, and in MicroPython display is used for hardware meanings (see examples) https://docs.micropython.org/en/latest/esp8266/tutorial/ssd1306.html

**Reply by fpliger** [reply]
> It’s not a builtin… your link doesn’t have a display builtin so it’s an obtrusive move to me as we have a module and asking Python developers to use a module is the normality, deciding ourselves what is builtin and what isn’t is not. MicroPython doesn’t have it neither, and neither does PythonMonkey. Why do we feel so special about what should be a builtin?

`PyScript != Python`. It's a Platform that allows languages (`Python` being the default, first and main one) in the browser with a simple interface.

Python has never run in the browser so it never had to even think about the topic. If it would, this discussion would have probably happened. PyScript vision and mission is not just to bring a language to the browser but to, among many others, provide a friendly and easy interface for people to program in a fun and easy way, reducing friction. 

> > Why we can't have already run from pyscript import display?

> if there is a display module users can import display as py_display and be more explicit, but I’m not sure I understood your question.

Sorry, I think I misread your answer early (I'm running around with a lot of distractions) when you wrote `display as global makes it impossible to have a display module`. 

In general, if `display` is global, yes, people should try to avoid naming things as `display` (while they still can). 

> display is already a pip module, that’s what I’m talking about: https://pypi.org/project/display/

Not sure I understand the point here. Yes, there's a `display` package name..... would you advise anyone to use that package?

> we can’t make up builtins, imho, and in MicroPython display is used for hardware meanings (see examples) https://docs.micropython.org/en/latest/esp8266/tutorial/ssd1306.html

Would that ever run in Micropython in pyscript?

---

In general, this is not a new thing. `display` is already in the the global namespace in PyScript. If that changes, Next would be taking it out.

**Reply by WebReflection** [reply]
PyScript Next supposed to be a new thing.

> Yes, there's a display package name..... would you advise anyone to use that package?

as a Python user I expect to be able to run any module pyodide offers. This one in particular is used by various projects including DB related and we have SQLite already.

I don’t know if pyodide offers this already but if it does, we’re being hostile, if it doesn’t, we’re consciously deciding to never want that in PyScript.

> Would that ever run in Micropython in pyscript?

@antocuni demoed PyScript in a Raspberry PI using USB or whatever, we have demoed Lego stuff and there could be displays in Lego too. PyScript runs via a browser but it’s not confined in browser only potentials.

In short, there is evidence this approach is obtrusive and no strong reasons or polls or users feedback in wanting us to make display a builtin. I don’t understand our decision making and the only evidence we have are:

  * there is a pip module named display **edit** I've checked and it's not available in pyodide: `Can't find a pure Python 3 wheel for 'display'.`
  * Users asked why we leak stuff globally
  * display is used for other things, not just show stuff on the main thread of a browser with a likely too little API / signature that doesn’t easily allow other non browser targets


What are the arguments to ignore all these points?

> If it would, this discussion would have probably happened.

In node GTK or any UI targeting project, including PyGTK AFAIK nobody decided non standard builtins so I keep being -1 on being hostile and not future-friendly by polluting with our own stuff otherwise fully compliant Python environment. I won’t oppose to it, but I’m not going to like it.

**Reply by WebReflection** [reply]
On the other hand, NodeJS and Bun offer process which is indeed a “builtin” runtime specific … so maybe we could compromise but I hope we can do that without breaking the pip module itself.

**amend** the pip module is apparently both not available in pyodide and not safe neither ... I stand corrected about this point.

**Comment by hoodmane** [comment]
In my opinion, if you DO want to add stuff to globals, it is worth considering adding them to `builtins`:
```py
import builtins
builtins.display = display
```
This ensures that they appear in all scopes, not just the main one.

**Reply by fpliger** [reply]
Good call :)

**Reply by WebReflection** [reply]
but what’s your take on doing that?

**Reply by WebReflection** [reply]
On a second thought, would exposing that as builtin allow usage of the `pip` display module whenever that would be desired? If yes is the answer then 2/3rd of my points are invalid and if it’s a builtin we’re after I fully agree builtins is what we should pollute.

**Reply by WebReflection** [reply]
checked and `Can't find a pure Python 3 wheel for 'display'.` so my issue is non existent ... I still wonder if adding builtins allow usage of modules with name clashing in case we want to add more than just `display` but I also believe I can test that myself, just wondering if there are implications, caveats or limits to consider.

**Reply by JeffersGlass** [reply]
If you had some module/package called `display` (from PYPI or local), and in a given namespace you called `import display`, it would shadow the included `display` function whether the included function is part of builtins or not. Adding it to builtins just makes it available as a name in any namespace (ie in other imported modules).

That said, if users are made aware of display, the workaround is to simply use `import display as disp` or similar, so as not to shadow the PyScript’s `display`. A similar case would be sometime like Rich’s `print` function, which they _intend_ to shadow the built-in print and give it extra functionality, but if you want to keep it separate you can do `from rich import print as rich_print` or similar.

**Reply by WebReflection** [reply]
Ok, thanks for explaining this fully. At this point though I’m not sure we should pollute builtins anymore, because that could break modules checking for it and doing something else if not available … as a global it feels less obtrusive in that regard, accordingly to your explanation, right?

**Reply by fpliger** [reply]
[The display package on PyPI](https://pypi.org/project/display/) is opaque at its best. I would not encourage anyone to install it as it seems to be missing much of the information needed for it to be considered a "safe" to install package. What's the code? Who maintains it? 

From the description, it's a `simple printer of nested list` and, in that case, I'd encourage people to use `pprint` instead.

This next release is a transitional release where many features have been removed [at least temporarily for most of them]. While I'm comfortable of changing the way we deal with `js`, `pyscript`, `Element`,  and (soft comfortable :) ) `HTML`, I am not with `display`. The risk of it being a mistake and having to revert back to how we did it is high and I'd muuuuuch rather allow enough time to think about this and discuss and eventually change it for a future release. If I look at `print` and the official Python `built-ins`, I cannot but think that if Python lived on the browser `display` would make a lot of sense as part of `built-ins`. 

I have similar feelings for `when` tbh (it feels very close to `property`, although I know they are quite different in terms of features they provide) but since `when` already had to be imported and there are alternatives, I agree that it makes total sense to not change it now (for the same reasons I mentioned above).

**Reply by WebReflection** [reply]
I think latest Antonio comment addresses all your desires and freedom to offer anything we like out of the `pyscript` module and *out of the box*, which is the cleanest approach to me, but I'd like to thank you for investigating that module internals ... to me that looks like a *pip* responsibility to grant publish modules are not "evil" but as I know pretty much nothing about the Python ecosystem I am happy to learn there could be modules nobody should use: thanks for digging that further!

**Comment by Neon22** [comment]
FWIW my opinion as a user is that I don't mind importing whatever as long as its documented what to do.
If you guys said you need to do `from pyscript import display, HTML, when` then that is what we'll all do.

I can see the arguments both ways and they're all legit. In my experience the only guiding principle would be  `You can't take it out once its in` (you can of course but its a real pain for user's existing code)

My qualifier is in a detail:
- `@when` works on initial instantiation but fails if an element is added from user interaction - such as resizing an array and adding some elements to support that. So IMHO  `@when` needs to work when elements are subsequently added to the dom . (implemented perhaps through the existing Observable in pydom but I am loathe to suggest implementation.)
 - for `py-events` if you wanted a test case then implementing a demo where DnD worked between two `divs`(dragging  an HTML element) and also between two `svgs` (dragging a `path`), would be an ideal test case because it would show how ugly the code needs to be to do something non-trivial.
 
Having pydom and building complex UIs from python is something I am really looking forward to.

**Reply by Neon22** [reply]
One more thing - as an early developer on micropython and in serving html for wireless connections from IOT hardware - I'd very much like micropython to offer similar UI functionality (when, html, py-event etc) as well as from py-iodide. Its very important IMHO.

**Reply by WebReflection** [reply]
FWIWI our display utility (pure Python) should work out of the box in MicroPython too. The gotcha with MicroPython in WASM is that we can't import packages at runtime, the whole thing needs to be compiled explicitly with "this or that" package, however you can use mostly everything we offer in PyScript in MicroPython too (beside pyodide packages) via [polyscript](https://github.com/pyscript/polyscript), if that helps somehow your experimentation. That is what fuels PyScript custom script type.

This might not answer your question or provide a solution out of the box, but maybe it helps you, and ourselves in the future, to provide what you are after for IoT related topics and use cases.

**Comment by WebReflection** [comment]
@fpliger @tedpatrick @mchilvers as discussed and as we want to release, it feels like the common agreed part around these topics is:

  1. we can bring `display` as builtin (I am still not super happy but it looks like most concerns I had are resolved)
  2. `when` was never leaked publicly, as Jeff mentioned, and `pydom` has a `@pydom.when` which I think is the right way to go
  3. *HTML* was a leak and both me, @JeffersGlass, and @Neon22 agree that if something need to be imported we can just state that in docs and while `display` has strong feelings around being the right builtin to expose, I am personally really skeptic about *HTML* or anything else we were leaking before (including `Element` which is already a DOM standard class). As we all know that once in we're doomed and can't remove it, can we move on with just `display` and provide a `pyweb` or `pydom` for DOM related things?
  4. if we agree on the 3rd point I am not sure what is the blocker here to have *pydom* or *pyweb* module integrated

Last, but not least, there are a lot of comments to address for PyDOM but none of these is related to this discussion. If there's anything else to clarify in there I'd be happy to do so but before releasing that MR needs to get out of draft and apply or discuss suggested changes.

**edit** I am fully OK with latest Antonio comment and resolution, so that all points except the MR little gotchas would be resolved by following Antonio suggestion.

**Comment by antocuni** [comment]
I'm just back from vacation. After reading the whole thread I seem to understand that there is no big consensus yet, so let me add my 2 cents.

I agree with @fpliger that `display()` is the "print for the web" and I would like it to be available "by default". If we decide to make it automatically available, +1 on @hoodmane suggestion to add it to the builtins.

However, putting `display` in the builtins creates additional problems:

- most python IDEs will not know about it: they will mark `display()` as a name error and most likely show it with a red underline
- same for linters & co.
- it makes it harder to google for docs: from the beginner's point of view, they don't know what comes "from python" (e.g. `print()`) and what comes "from pyscript" (e.g. `display()`). As of today, if you google `python display` you get documentation for IPython, if you google `pyscript display` you get the right pyscript docs. If they see `from pyscript import display` it makes it very clear that it's part of pyscript.
- The previous point is something which I actually saw in action when I looked at TUFTS students: they were very surprised that `display` came from nowhere.

I have an alternative suggestion: we could just provide a file template which does `from pyscript import display, HTML, when` by default.
I expect that a big majority of "beginners" will use pyscript via pyscript.com: so, if pyscript.com just adds that line at the beginning of every new `.py` file that the user creates, we solve the problem for the 90% of people.

People using pyscript *outside* of pyscript.com are very likely more experienced, and for *them* having an explicit import is surely less confusing than an implicit one.

**Reply by WebReflection** [reply]
I am +1 with everything Antonio said here and indeed it uses what I previously suggested which, if automated, doesn't look like creating friction or even confusion, as anyone can look at the source code of the page/file and also spot that with ease.

Doing this way, we'd be free to add any convenient global/builtin utility as we see fit without ever being too obtrusive for more advanced users: it's (imho) the cleanest way to go.

**Reply by antocuni** [reply]
and we can automate it even more with the `pyscript` CLI command.
`pyscript create` would create a new project which consist of:
- an `index.html` template with the proper `<script>` and `<link>` tags
- a `main.py` template with `from pyscript import ...`
- possibly a `config.toml`, exactly as we do for the .com template

**Reply by fpliger** [reply]
This is a compelling point. Let's try to converge on this as a group tomorrow but I'm not against it (While I still really believe that `display` is the `print` for the web, your additional problems around putting it as a builtin make sense, especially around the reusability of examples and IDE integration if it's "just a Python file". At least today.)

**Reply by Neon22** [reply]
IWBNI we could choose between a regular template and a worker enhanced template. Just asking.

**Reply by fpliger** [reply]
Could be a scenario, yes.

Ideally, it'd be nice to not need that because, ideally, it be nicer if a "regular" template is just like a `worker`. At this point, I think we need to understand what the use cases are and how to better provide these templates to users... 

With that said, my 2c is that we'll learn that, even if we will try to make the difference between main thread vs. workers transparent, workers will have some aspects very specific to them that will make sense to create templates dedicated to them (even just from a optimization/supported features point of view)

**Reply by Neon22** [reply]
This is possibly the wrong place for this but my `worker` usecase is:
- in pyweaving - send to a `worker` a job to do a time consuming itertools search to find all optimal solutions to a minimization problem.
- while the main thread dealt purely with UI response and building canvas svg elements.
- there are a number of tasks like this that I would dispatch to workers, then reintegrate into the UI when the task finished.

I suspect that this is a typical Workers usecase - along with async fetch requests for remote data.

**Reply by WebReflection** [reply]
FYI current worker supports `display` transparently (also same import from pyscript) and it has full DOM access too so it's possible that you can just write code as if it was in the main thread but in a non-blocking way via worker ... use `window` (instead of `js`) and `document` (instead of `js.document`) from `pyscript` and see if that's enough.
